core/src/main/scala/org/apache/spark/RecoverCacheShutdown.scala

```diff
@@ -0,0 +1,234 @@
+        val blocksRemain = blocksToSave.get(id).exists(_.nonEmpty)
+        id -> (if (blocksRemain) HasCachedBlocks else NoMoreBlocks)
+      } else {
+        id -> NotEnoughMemory
+      }
+    }(collection.breakOut)
+  }
+
+  /**
+   *  Is there enough memory on the cluster to replicate the cached data on execId before we delete
+   *  it. Take into account all the blocks that are currently on track to be deleted.
+   * @param execId the id of the executor we want to delete.
+   * @return true if there is room
+   */
+  private def isThereEnoughMemory(execId: String): Boolean = {
+    val currentMemStatus = blockManagerMasterEndpoint
+      .askSync[Map[BlockManagerId, (Long, Long)]](GetMemoryStatus)
+      .map { case (blockManagerId, mem) => blockManagerId.executorId -> mem }
+    val remaining = currentMemStatus - execId -- blocksToSave.keys
+
+    val thisExecBytes = if (!blocksToSave.contains(execId) && currentMemStatus.contains(execId)) {
+      currentMemStatus(execId)._1 - currentMemStatus(execId)._2
+    } else {
+      0
+    }
+
+    remaining.nonEmpty && everyExecutorHasEnoughMemory(remaining, thisExecBytes)
+  }
+
+  private def everyExecutorHasEnoughMemory(execMem: Map[String, (Long, Long)], bytes: Long) = {
+    val blocksInQueueBytes = blockManagerMasterEndpoint.askSync[Long](GetSizeOfBlocks(blockList))
+    val bytesPerExec = (blocksInQueueBytes + bytes) / execMem.size.toFloat
+    execMem.forall { case (_, (_, remaining)) => remaining - bytesPerExec >= 0 }
+  }
+
+  private def blockList = for {
+    (id, queue) <- blocksToSave.toSeq
+    block <- queue
+  } yield (id, block)
+
+  private def updateBlockState(execId: String): Unit = {
+    val blocks: mutable.Set[RDDBlockId] = blockManagerMasterEndpoint
+      .askSync[collection.Set[BlockId]](GetCachedBlocks(execId))
+      .flatMap(_.asRDDId)(collection.breakOut)
+
+    blocks --= savedBlocks.getOrElse(execId, new mutable.HashSet)
+    blocksToSave(execId) = mutable.PriorityQueue[RDDBlockId](blocks.toSeq: _*)(Ordering.by(_.rddId))
+    killTimers.getOrElseUpdate(execId, scheduleKill(new KillRunner(execId)))
+  }
+
+  class KillRunner(execId: String) extends Runnable { def run(): Unit = killExecutor(execId) }
+  private def scheduleKill(k: KillRunner) = scheduler.schedule(k, forceKillAfterS, TimeUnit.SECONDS)
+
+  /**
+   * Ask block manager master to replicate one cached block.
+   * @param execId the executor to replicate a block from
+   * @return false if there are no more blocks or if there is an issue
+   */
+  def replicateFirstBlock(execId: String): (Future[Boolean], Option[RDDBlockId]) = synchronized {
+    logDebug(s"Replicate block on executor $execId.")
+    blocksToSave.get(execId) match {
+      case Some(p) if p.nonEmpty => doReplication(execId, p.dequeue())
+      case _ => (Future.successful(false), None)
+    }
+  }
+
+  private def doReplication(execId: String,
+                            blockId: RDDBlockId): (Future[Boolean], Option[RDDBlockId]) = {
```

Comment:
nit: this is not how we do multi-line arg lists.

Reply:
Fixed.